OpenTelemetry Runtime Metrics - Update landing page to consolidate information (#28657)
DataDog/documentation: content/en/opentelemetry/integrations/runtime_metrics/_index.md

title: OpenTelemetry Runtime Metrics
aliases:
- /opentelemetry/runtime_metrics/
- /opentelemetry/integrations/runtime_metrics/go/
- /opentelemetry/integrations/runtime_metrics/dotnet/
- /opentelemetry/integrations/runtime_metrics/java/
further_reading:
- link: "/tracing/metrics/runtime_metrics/"
  tag: "Documentation"
  text: "APM Runtime Metrics"
- link: "/opentelemetry/mapping/metrics_mapping/"
  tag: "Documentation"
  text: "OpenTelemetry Metrics Mapping"
---

## Overview

Runtime metrics provide insights into application performance, including memory usage, garbage collection, and parallelization. Datadog tracing libraries offer [runtime metrics collection][5] for each supported language, and OpenTelemetry (OTel) also collects compatible runtime metrics that can be sent to Datadog through the OpenTelemetry SDKs.

## Compatibility

Datadog supports OpenTelemetry runtime metrics for the following languages:
- Java
- .NET
- Go

For details about host and container metrics mapping, see [OpenTelemetry Metrics Mapping][1].

## Setup instructions

### 1. Prerequisites

- You have successfully [configured OpenTelemetry metrics to send to Datadog][2].

> - You have installed the corresponding language integration.

Reviewer: Is this talking about the SDK? I think it would be helpful to link to what we are talking about here.

Author: This is referring to the integration _in the Datadog app_, which comes with an OOTB dashboard. Do you have any opinions on how we can add a link for the specific programming language they're interested in?

Reviewer: Ah! We use "integrations" for so many things I wasn't sure.

Maybe we could just link to https://app.datadoghq.com/integrations to make it extra clear?

Author: Updated, let me know what you think!

Reviewer: Thanks! Looks good to me.

#### Automatic instrumentation

If you use [OpenTelemetry automatic instrumentation][3] for .NET applications, runtime metrics are enabled by default.

#### Manual instrumentation

If you use [OpenTelemetry manual instrumentation][4], see the documentation for the [OpenTelemetry.Instrumentation.Runtime library][5].

#### Metric export interval

The default metric export interval for the .NET OTel SDK differs from the default for the Datadog .NET SDK. Datadog recommends setting the [OTEL_METRIC_EXPORT_INTERVAL][7] environment variable on your .NET service to match the default Datadog metric export interval:

```
OTEL_METRIC_EXPORT_INTERVAL=10000
```

[3]: https://opentelemetry.io/docs/instrumentation/net/automatic/
[4]: https://opentelemetry.io/docs/instrumentation/net/manual/
[5]: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.Runtime
[7]: https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#periodic-exporting-metricreader

{{% /tab %}}

{{< /tabs >}}

## View runtime metric dashboards

After setup is complete, you can view runtime metrics in:

> - The service's details page (see Java example below)

Reviewer: This screenshot is outdated. We should find a new one for this example.

Author: Updated in df4e9862c7b763f831f2304f2294fae2ce4c699b

[4]: https://opentelemetry.io/docs/instrumentation/net/manual/
[5]: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.Runtime
[7]: https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#periodic-exporting-metricreader

</div>

{{% /tab %}}

{{< /tabs >}}

## View runtime metric dashboards

After setup is complete, you can view runtime metrics in:
- The service's details page (see Java example below)
- The flame graph metrics tab
- Default [runtime dashboards][7]

{{< img src="opentelemetry/otel_runtime_metrics_service_page.png" alt="Service page showing OpenTelemetry runtime metrics on the JVM Metrics tab" style="width:100%;" >}}

## Data collected

When using OpenTelemetry runtime metrics with Datadog, you receive both:
- Original OpenTelemetry runtime metrics
- Mapped Datadog runtime metrics for equivalent metrics

The OpenTelemetry runtime metrics have the following prefixes based on their source:

| Source | Prefix |
| --- | --- |
| [OTel Collector Datadog Exporter][100] | `otel.process.runtime.*` |
| [Datadog Agent OTLP Ingest][101] | `process.runtime.*` |

> The following tables list the Datadog runtime metrics that are supported through OpenTelemetry mapping. "N/A" indicates that there is no OpenTelemetry counterpart.

Reviewer: ```suggestion
The following tables list the Datadog runtime metrics that are supported through OpenTelemetry mapping. "N/A" indicates that there is no OpenTelemetry equivalent metric available.
```

<div class="alert alert-warning"> OpenTelemetry runtime metrics are mapped to Datadog by metric name. Do not rename host metrics for OpenTelemetry runtime metrics as this breaks the mapping.</div>

[100]: /opentelemetry/setup/collector_exporter/
[101]: /opentelemetry/setup/otlp_ingest_in_the_agent

{{< tabs >}}
{{% tab "Java" %}}

> | Datadog metric | Description |  OpenTelemetry counterpart |

Reviewer: ```suggestion
| Datadog metric | Description |  OpenTelemetry metric |
```

| `jvm.buffer_pool.mapped.count` | Number of mapped buffers in the pool. | `process.runtime.jvm.buffer.count`<br> `jvm.buffer.count` |
| `jvm.buffer_pool.mapped.limit` | Measure of total memory capacity of mapped buffers. | `process.runtime.jvm.buffer.limit` <br> `jvm.buffer.memory.limit` |
| `jvm.gc.parnew.time` | The approximate accumulated garbage collection time elapsed. | N/A |
| `jvm.gc.cms.count` | The total number of garbage collections that have occurred. | N/A |
| `jvm.gc.major_collection_count` | The rate of major garbage collections. Set `new_gc_metrics: true` to receive this metric. | N/A |
| `jvm.gc.minor_collection_count` | The rate of minor garbage collections. Set `new_gc_metrics: true` to receive this metric. | N/A |
| `jvm.gc.major_collection_time` | The fraction of time spent in major garbage collection. Set `new_gc_metrics: true` to receive this metric. | N/A |
| `jvm.gc.minor_collection_time` | The fraction of time spent in minor garbage collection. Set `new_gc_metrics: true` to receive this metric. | N/A |
| `jvm.os.open_file_descriptors` | The number of open file descriptors. | N/A |

{{% /tab %}}

{{% tab "Go" %}}

> | Datadog metric | Description |  OpenTelemetry counterpart |

Reviewer: ```suggestion
| Datadog metric | Description |  OpenTelemetry metric |
```

| `runtime.go.mem_stats.gc_sys` | Bytes of memory in garbage collection metadata. | N/A |
| `runtime.go.mem_stats.other_sys` | Bytes of memory in miscellaneous off-heap. | N/A |
| `runtime.go.mem_stats.next_gc` | Target heap size of the next GC cycle. | N/A |
| `runtime.go.mem_stats.last_gc` | Last garbage collection finished, as nanoseconds since the UNIX epoch. | N/A |
| `runtime.go.mem_stats.num_forced_gc` | Number of GC cycles that were forced by the application calling the GC function. | N/A |
| `runtime.go.mem_stats.gc_cpu_fraction` | Fraction of this program's available CPU time used by the GC since the program started. | N/A |
| `runtime.go.gc_stats.pause_quantiles.min` | Distribution of GC pause times: minimum values. | N/A |
| `runtime.go.gc_stats.pause_quantiles.25p` | Distribution of GC pause times: 25th percentile. | N/A |
| `runtime.go.gc_stats.pause_quantiles.50p` | Distribution of GC pause times: 50th percentile. | N/A |
| `runtime.go.gc_stats.pause_quantiles.75p` | Distribution of GC pause times: 75th percentile. | N/A |
| `runtime.go.gc_stats.pause_quantiles.max` | Distribution of GC pause times: maximum values. | N/A |

{{% /tab %}}

{{% tab ".NET" %}}

> | Datadog metric | Description |  OpenTelemetry counterpart |

Reviewer: ```suggestion
| Datadog metric | Description |  OpenTelemetry metric |
```

If you have instrumented your Java application with [OpenTelemetry manual instrumentation][4], see the following guides for your Java version:
- [Java 8][5]
- [Java 17][6]

[3]: https://opentelemetry.io/docs/instrumentation/java/automatic/
[4]: https://opentelemetry.io/docs/instrumentation/java/manual/
[5]: https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/runtime-telemetry/runtime-telemetry-java8/library
[6]: https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/runtime-telemetry/runtime-telemetry-java17/library

{{% /tab %}}

{{% tab "Go" %}}

OpenTelemetry (OTel) Go applications are [instrumented manually][3]. To enable runtime metrics, see the documentation for the [runtime package][4].

[3]: https://opentelemetry.io/docs/instrumentation/go/manual/
[4]: https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/runtime

{{% /tab %}}

{{% tab ".NET" %}}

<div class="alert alert-warning">The minimum supported version of the .NET OpenTelemetry SDK is <a href="https://github.com/open-telemetry/opentelemetry-dotnet/releases/tag/core-1.5.0">1.5.0</a></div>

> If you have instrumented your .NET applications with [OpenTelemetry automatic instrumentation][3], runtime metrics are automatically enabled. 

Reviewer 2: This is probably out of the scope of this PR, but the .NET section likely needs to be updated since the runtime metric names have changed starting .NET 9: https://learn.microsoft.com/en-us/dotnet/core/diagnostics/built-in-metrics-runtime.

- `OTEL_METRIC_EXPORT_INTERVAL=10000`

[3]: https://opentelemetry.io/docs/instrumentation/net/automatic/
[4]: https://opentelemetry.io/docs/instrumentation/net/manual/
[5]: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.Runtime
[7]: https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#periodic-exporting-metricreader

{{% /tab %}}

{{< /tabs >}}

## View runtime metric dashboards

After setup is complete, see your runtime metrics in the service's details page (see Java example below), the flame graph metrics tab, and in [default runtime dashboards][7].

{{< img src="opentelemetry/otel_runtime_metrics_service_page.png" alt="Service page showing OpenTelemetry runtime metrics on the JVM Metrics tab" style="width:100%;" >}}

## Data collected

When using OpenTelemetry runtime metrics with Datadog, you receive both the original OpenTelemetry runtime metrics as well as mapped Datadog runtime metrics for equivalent metrics. The OpenTelemetry runtime metrics have the following prefixes based on their source:

> | OTel Collector Datadog Exporter | Datadog Agent OTLP Ingest |

Reviewer 2: Not sure the best way to format this, but would be good to include DDOT (Datadog Agent w/ OpenTelemetry Collector) in this chart.
```suggestion
| OTel Collector Datadog Exporter | Datadog Agent OTLP Ingest / DDOT |
```

Reviewer 2: Sure that works
